src/nvim/regexp.c

```diff
@@ -6287,43 +6287,38 @@ static int cstrncmp(char_u *s1, char_u *s2, int *n)
 /*
  * cstrchr: This function is used a lot for simple searches, keep it fast!
  */
-static char_u *cstrchr(char_u *s, int c)
+static inline char_u *cstrchr(const char_u *const s, const int c)
+  FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_ALL
+  FUNC_ATTR_ALWAYS_INLINE
 {
-  char_u      *p;
-  int cc;
-
-  if (!ireg_ic
-      || (!enc_utf8 && mb_char2len(c) > 1)
-      )
+  if (!ireg_ic) {
     return vim_strchr(s, c);
+  }
+
+  // tolower() and toupper() can be slow, comparing twice should be a lot
+  // faster (esp. when using MS Visual C++!).
+  // For UTF-8 need to use folded case.
+  if (c > 0x80) {
+    const int folded_c = utf_fold(c);
+    for (const char_u *p = s; *p != NUL; p += utfc_ptr2len(p)) {
+      if (utf_fold(utf_ptr2char(p)) == folded_c) {
+        return (char_u *)p;
+      }
+    }
+    return NULL;
+  }
```

Comment:
You're right, I got confused by the diff reordering. I thought this part was new.